atc/db/team.go

```diff
@@ -1567,19 +1499,158 @@ func inactivateTableForPipeline(tx Tx, pipelineID int, tableName string) error {
 }
 
 func saveResources(tx Tx, resources atc.ResourceConfigs, pipelineID int) (map[string]int, error) {
-	resourceNameToID := make(map[string]int)
+	if len(resources) == 0 {
+		return nil, nil
+	}
+	insertQuery := psql.Insert("resources").
+		Columns("name", "pipeline_id", "config", "active", "nonce", "type", "resource_config_id", "resource_config_scope_id").
+		Suffix("ON CONFLICT (name, pipeline_id) DO UPDATE SET config = EXCLUDED.config, active = EXCLUDED.active, nonce = EXCLUDED.nonce, type = EXCLUDED.type, resource_config_id = EXCLUDED.resource_config_id, resource_config_scope_id = EXCLUDED.resource_config_scope_id").
+		Suffix("RETURNING name, id")
+	resourcesToPin := map[string][]byte{}
+
+	existingResources, err := existingResources(tx, pipelineID)
+	if err != nil {
+		return nil, nil
+	}
+
 	for _, resource := range resources {
-		resourceID, err := saveResource(tx, resource, pipelineID)
+		configPayload, err := json.Marshal(resource)
 		if err != nil {
 			return nil, err
 		}
 
-		resourceNameToID[resource.Name] = resourceID
+		es := tx.EncryptionStrategy()
+		encryptedPayload, nonce, err := es.Encrypt(configPayload)
+		if err != nil {
+			return nil, err
+		}
+		values := []interface{}{resource.Name, pipelineID, encryptedPayload, true, nonce, resource.Type}
+
+		existing, exists := existingResources[resource.Name]
+		if !exists || (mapHash(resource.Source) != mapHash(existing.Config.Source) || resource.Type != existing.Type) {
+			values = append(values, nil, nil)
+		} else {
+			values = append(values, existing.ResourceConfigID, existing.ResourceConfigScopeID)
+		}
+
+		if resource.Version != nil {
+			version, err := json.Marshal(resource.Version)
+			if err != nil {
+				return nil, err
+			}
+			resourcesToPin[resource.Name] = version
+		}
+
+		insertQuery = insertQuery.Values(values...)
 	}
 
+	resourceNameToID := make(map[string]int)
+	resourceIDs := []int{}
+
+	rows, err := insertQuery.RunWith(tx).Query()
+	if err != nil {
+		return nil, err
+	}
+	for rows.Next() {
+		var (
+			name string
+			id   int
+		)
+		rows.Scan(&name, &id)
+		resourceNameToID[name] = id
+		resourceIDs = append(resourceIDs, id)
+	}
+	err = resetResourcePins(tx, resourceIDs, resourcesToPin, resourceNameToID)
+	if err != nil {
+		return nil, err
+	}
 	return resourceNameToID, nil
 }
 
+type existingResource struct {
+	Type                  string
+	Config                atc.ResourceConfig
+	ResourceConfigID      interface{}
+	ResourceConfigScopeID interface{}
```

Comment:
These only exist to be used in a squirrel query `.Values()` function so the type isn't that relevant - the main thing is we want them to be nil-able (rather than 0 which is the default for `int`)

Reply:
Is there a reason why you want to avoid doing a `.Valid` check on them?